pkg/plugins/balance/balance.go

```diff
@@ -0,0 +1,71 @@
+package balance
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/labels"
+	clusterapiv1 "open-cluster-management.io/api/cluster/v1"
+	clusterapiv1alpha1 "open-cluster-management.io/api/cluster/v1alpha1"
+	"open-cluster-management.io/placement/pkg/plugins"
+)
+
+const (
+	placementLabel = "cluster.open-cluster-management.io/placement"
+	description    = `
+	Balance prioritizer balance the number of decisions among the clusters. The cluster
+	with the highest number of decison is given the lowest score, while the empty cluster is given
+	the highest score.
+	`
+)
+
+var _ plugins.Prioritizer = &Balance{}
+
+type Balance struct {
+	handle plugins.Handle
+}
+
+func New(handle plugins.Handle) *Balance {
+	return &Balance{handle: handle}
+}
+
+func (b *Balance) Name() string {
+	return "balance"
+}
+
+func (b *Balance) Description() string {
+	return description
+}
+
+func (b *Balance) Score(ctx context.Context, placement *clusterapiv1alpha1.Placement, clusters []*clusterapiv1.ManagedCluster) (map[string]int64, error) {
+	scores := map[string]int64{}
+	for _, cluster := range clusters {
+		scores[cluster.Name] = plugins.MaxClusterScore
+	}
+
+	decisions, err := b.handle.DecisionLister().List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	var maxCount int64
+	decisionCount := map[string]int64{}
+	for _, decision := range decisions {
+		// Do not count the decision that is being scheduled.
+		if decision.Labels[placementLabel] == placement.Name && decision.Namespace == placement.Namespace {
+			continue
+		}
+		for _, d := range decision.Status.Decisions {
+			decisionCount[d.ClusterName] = decisionCount[d.ClusterName] + 1
+			if decisionCount[d.ClusterName] > maxCount {
+				maxCount = decisionCount[d.ClusterName]
+			}
+		}
+	}
+
+	for clusterName := range scores {
+		if count, ok := decisionCount[clusterName]; ok {
+			scores[clusterName] = scores[clusterName] - scores[clusterName]*count/maxCount
```

Comment:
I'm surprised this works with integer mat.  I would expect count to be less than maxCount for nearly everything, zero-ing out the subtraction.  I think the output score should be integer because it's human friendly, but the math leading to it may need to be floating point.

Reply:
for instance,

1. compute percentage for how loaded they are relative to max
2. subtract 0.5 (this centers us on zero)
4. negate the value (this make us prefer those which are less laden)
5. multiply by 200 (since we ranged from -0.5 to 0.5, this give us +/-100)

Reply:
current math is to calculate the free with 100 * (1 - count/max_count). I think the reason to make it +/-100 is to imply that "I do not want to select these clusters" while if range is from 0/100 will not have such implication? so it will be like 200 * (0.5 - count/max_count)